agent/provider/src/payments/payments.rs

```diff
@@ -494,82 +501,106 @@ impl Handler<UpdateCost> for Payments {
+        {
+            Ok(agreement) => agreement,
+            Err(e) => return ActorResponse::reply(Err(e)),
         };
 
-        if let Some(activity) = agreement.activities.get(&msg.invoice_info.activity_id) {
-            if let ActivityPayment::Running { .. } = activity {
+        return match agreement.activities.get(&msg.invoice_info.activity_id) {
+            Some(ActivityPayment::Running { .. }) => {
                 let payment_model = agreement.payment_model.clone();
                 let context = self.context.clone();
+
+                let update_interval = agreement.update_interval;
+                let last_debit_note = agreement.last_send_debit_note.clone();
+                let deadline = agreement.payment_deadline.clone();
                 let invoice_info = DebitNoteInfo {
-                    payment_deadline: agreement.payment_deadline.clone(),
+                    payment_deadline: deadline.clone(),
                     ..msg.invoice_info.clone()
                 };
-                let update_interval = agreement.update_interval;
 
                 let debit_note_future = async move {
                     let (debit_note, _cost) = compute_cost_and_send_debit_note(
                         context.clone(),
                         payment_model.clone(),
                         &invoice_info,
                     )
-                    .await?;
+                    .await
+                    .log_err()?;
                     Ok(debit_note)
                 }
                 .into_actor(self)
-                .map(move |result: Result<DebitNote, Error>, _, ctx| {
-                    if let Err(error) = result {
-                        log::error!("{}", error)
+                .map(move |result: Result<_, anyhow::Error>, myself, ctx| {
+                    // We break Agreement, if we weren't able to send any DebitNote lately.
+                    if result.is_err() {
+                        if deadline.is_some() && Utc::now() > last_debit_note + deadline.unwrap() {
+                            myself.break_agreement_signal
+                                .send_signal(BreakAgreement {
+                                    agreement_id: msg.invoice_info.agreement_id.clone(),
+                                    reason: BreakReason::RequestorUnreachable(deadline.unwrap()),
+                                })
+                                .log_err_msg(&format!(
+                                    "Failed to send BreakAgreement for [{}], when Requestor is unreachable.",
+                                    msg.invoice_info.agreement_id
+                                ))
+                                .ok();
+                        }
+                    } else {
+                        myself.agreements
+                            .get_mut(&msg.invoice_info.agreement_id)
+                            .map(|agreement| agreement.last_send_debit_note = Utc::now());
                     }
+
```

Comment:
Is the below comment and action still valid? Shouldn't we rather withhold sending debit notes when `BreakAgreeement` was already sent?

Reply:
I could do it, but it will happen anyway